src/databricks/labs/ucx/hive_metastore/tables.py

```diff
@@ -220,21 +215,114 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
             return None
 
 
+@dataclass
+class MigrationCount:
+    database: str
+    managed_tables: int = 0
+    external_tables: int = 0
+    views: int = 0
+
+
+class TableMigrationStrategy:
+    def __init__(
+        self,
+        ws: WorkspaceClient,
+        backend: SqlBackend,
+        source_table: Table,
+        target_table_key: str,
+    ):
+        self._backend = backend
+        self._ws = ws
+        self._source_table = source_table
+        self._target_table_key = target_table_key
+
+    @abstractmethod
+    def migrate_object(self):
+        raise NotImplementedError
+
+
+class ExternalTableTargetStrategy(TableMigrationStrategy):
+    def __init__(
+        self,
+        ws: WorkspaceClient,
+        backend: SqlBackend,
+        source_table: Table,
+        target_table_key: str,
+    ):
+        super().__init__(ws, backend, source_table, target_table_key)
+
+    def migrate_object(self):
+        source_table = self._source_table
+        target_table_key = self._target_table_key
+        sql = source_table.sql_external(target_table_key)
+        logger.debug(f"Migrating table {source_table.key} to {target_table_key} (EXTERNAL) using SQL query: {sql}")
+        self._backend.execute(sql)
+
+
+class DBFSRootToManagedStrategy(TableMigrationStrategy):
+    def __init__(
+        self,
+        ws: WorkspaceClient,
+        backend: SqlBackend,
+        source_table: Table,
+        target_table_key: str,
+    ):
+        super().__init__(ws, backend, source_table, target_table_key)
+
+    def migrate_object(self):
+        source_table = self._source_table
+        target_table_key = self._target_table_key
+        sql = source_table.sql_managed(target_table_key)
+        logger.debug(
+            f"Migrating table {source_table.key} to {target_table_key} (DBFS Root to Managed) using SQL query: {sql}"
```

Comment:
Everywhere a table name is converted to a string, the table name should be encapsulated in `